Add TF ResNet model

amyeroberts opened the pull request:
Adds a tensorflow implementation of the ResNet model + associated tests.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

amyeroberts committed:
Rought TF conversion outline

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17427). All of your documentation changes will be reflected on that endpoint.

amyeroberts committed:
Tidy up

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Fix padding differences between layers

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Resolve conflicts

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Add back embedder - whoops

amyeroberts committed:
Merge in main

amyeroberts committed:
Match test file to main

amyeroberts committed:
Match upstream test file

sayakpaul commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,542 @@
+        return {**base_config, **config}
+
+
+@add_start_docstrings(
+    "The bare ResNet model outputting raw features without any specific head on top.",
+    RESNET_START_DOCSTRING,
+)
```
@amyeroberts we usually also add a `...MainLayer` (which should also be `keras_serializable`) where the components are collated. This main layer is then called from the downstream classes like `...ForImageClassification`, `TFXXXModel`, and so on. See here for an example:

https://github.com/huggingface/transformers/blob/224bde91caff4ccfd12277ab5e9bf97c61e22ee9/src/transformers/models/convnext/modeling_tf_convnext.py#L275

amyeroberts replied:
Thanks for pointing this out - I wasn't aware. I'll rework it.

amyeroberts replied:
Done!

sayakpaul commented on src/transformers/models/resnet/configuration_resnet.py:
```diff
@@ -95,6 +98,7 @@ def __init__(
         self.layer_type = layer_type
         self.hidden_act = hidden_act
         self.downsample_in_first_stage = downsample_in_first_stage
+        self.image_size = 224
```
self.image_size = image_size

amyeroberts replied:
Done!

sayakpaul commented on tests/models/resnet/test_modeling_tf_resnet.py:
```diff
@@ -0,0 +1,247 @@
+
+    def get_config(self):
+        return ResNetConfig(
+            num_channels=self.num_channels,
```
Shouldn't we also pass `image_size` here?

amyeroberts replied:
Yep! Fixed now

amyeroberts commented on src/transformers/models/resnet/configuration_resnet.py:
```diff
@@ -95,6 +98,7 @@ def __init__(
         self.layer_type = layer_type
         self.hidden_act = hidden_act
         self.downsample_in_first_stage = downsample_in_first_stage
+        self.image_size = 224
```
```suggestion
        self.image_size = image_size
```

amyeroberts commented on tests/models/resnet/test_modeling_tf_resnet.py:
```diff
@@ -0,0 +1,247 @@
+            hidden_sizes=self.hidden_sizes,
+            depths=self.depths,
+            hidden_act=self.hidden_act,
+            num_labels=self.num_labels,
```
```suggestion
            num_labels=self.num_labels,
            image_size=self.image_size,
```

amyeroberts committed:
Correctly pass and assign image_size parameter

Co-authored-by: Sayak Paul <spsayakpaul@gmail.com>

amyeroberts committed:
Add in MainLayer

amyeroberts committed:
Correctly name layer

amyeroberts commented on tests/models/resnet/test_modeling_tf_resnet.py:
```diff
@@ -0,0 +1,252 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Testing suite for the PyTorch ResNet model. """
```
```suggestion
""" Testing suite for the Tensorflow ResNet model. """
```

amyeroberts replied:
```suggestion
""" Testing suite for the Tensorflow ResNet model. """
```

amyeroberts commented on src/transformers/modeling_tf_outputs.py:
```diff
@@ -46,6 +46,25 @@ class TFBaseModelOutput(ModelOutput):
+    """
+
+    last_hidden_state: tf.Tensor = None
+    hidden_states: Optional[Tuple[tf.Tensor]] = None
```
```suggestion
    hidden_states: Optional[Tuple[tf.Tensor, ...]] = None
```

amyeroberts commented on src/transformers/modeling_tf_outputs.py:
```diff
@@ -80,6 +99,28 @@ class TFBaseModelOutputWithPooling(ModelOutput):
+
+    last_hidden_state: tf.Tensor = None
+    pooler_output: tf.Tensor = None
+    hidden_states: Optional[Tuple[tf.Tensor]] = None
```
```suggestion
    hidden_states: Optional[Tuple[tf.Tensor, ...]] = None
```

amyeroberts commented on src/transformers/modeling_tf_outputs.py:
```diff
@@ -825,3 +866,24 @@ class TFSequenceClassifierOutputWithPast(ModelOutput):
+
+    loss: Optional[tf.Tensor] = None
+    logits: tf.Tensor = None
+    hidden_states: Optional[Tuple[tf.Tensor]] = None
```
```suggestion
    hidden_states: Optional[Tuple[tf.Tensor, ...]] = None
```

amyeroberts commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,572 @@
+
+    def call(self, hidden_state: tf.Tensor, training: bool = False) -> tf.Tensor:
+        residual = hidden_state
+
```
```suggestion
```

amyeroberts commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,572 @@
+        hidden_state = self.conv0(hidden_state, training=training)
+        hidden_state = self.conv1(hidden_state, training=training)
+        hidden_state = self.conv2(hidden_state, training=training)
+
```
```suggestion
```

amyeroberts committed:
Tidy up AdaptivePooler

amyeroberts committed:
Small tidy-up

More accurate type hints and remove whitespaces

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Merge branch 'add-tf-resnet' of github.com:amyeroberts/transformers into add-tf-resnet

amyeroberts committed:
Change AdaptiveAvgPool

Use the AdaptiveAvgPool implementation by @Rocketknight1, which correctly pools if the output shape does not evenly divide by input shape c.f. https://github.com/huggingface/transformers/pull/17554/files/9e26607e22aa8d069c86b50196656012ff0ce62a#r900109509

Co-authored-by: From: matt <rocketknight1@gmail.com>
Co-authored-by: Sayak Paul <spsayakpaul@gmail.com>

amyeroberts committed:
Use updated AdaptiveAvgPool

Co-authored-by: matt <rocketknight1@gmail.com>

amyeroberts committed:
Make AdaptiveAvgPool compatible with CPU

amyeroberts committed:
Remove image_size from configuration

amyeroberts committed:
Fixup

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

sayakpaul commented on docs/source/en/model_doc/resnet.mdx:
```diff
@@ -31,7 +31,7 @@ The figure below illustrates the architecture of ResNet. Taken from the [origina
 <img width="600" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/resnet_architecture.png"/>
 
-This model was contributed by [Francesco](https://huggingface.co/Francesco). The original code can be found [here](https://github.com/KaimingHe/deep-residual-networks).
+This model was contributed by [Francesco](https://huggingface.co/Francesco). The Tensorflow version of this model was added by [amyeroberts](https://huggingface.co/amyeroberts). The original code can be found [here](https://github.com/KaimingHe/deep-residual-networks).
```
Small nit: TensorFlow.

amyeroberts replied:
```suggestion
This model was contributed by [Francesco](https://huggingface.co/Francesco). The TensorFlow version of this model was added by [amyeroberts](https://huggingface.co/amyeroberts). The original code can be found [here](https://github.com/KaimingHe/deep-residual-networks).
```

amyeroberts committed:
Tensorflow -> TensorFlow

amyeroberts commented:
Swapping @sgugger and @LysandreJik as Lysandre's off and adding @Rocketknight1 for the TF side.

amyeroberts committed:
Fix pt references in tests

NielsRogge commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+_CHECKPOINT_FOR_DOC = "microsoft/resnet-50"
+_EXPECTED_OUTPUT_SHAPE = [1, 2048, 7, 7]
+
+# Image classification docstring
+_IMAGE_CLASS_CHECKPOINT = "microsoft/resnet-50"
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tiger cat"
```
Can you also add TFResNet to the doc tests?

NielsRogge commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+
+class TFResNetBasicLayer(tf.keras.layers.Layer):
+    """
+    A classic ResNet's residual layer composed by a two `3x3` convolutions.
```
```suggestion
    A classic ResNet's residual layer composed by two `3x3` convolutions.
```

NielsRogge commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+
+class TFResNetBottleNeckLayer(tf.keras.layers.Layer):
+    """
+    A classic ResNet's bottleneck layer composed by a three `3x3` convolutions.
```
```suggestion
    A classic ResNet's bottleneck layer composed by three `3x3` convolutions.
```

NielsRogge commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+"""
+
+
+# Copied from:
```
I'm afraid that if you add a Copied from statement like this, it might mess up with make fix-copies.

I usually use something like "taken from" in that case.

amyeroberts replied:
```suggestion
# Taken from:
```

amyeroberts committed:
Apply suggestions from code review - grammar and wording

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

amyeroberts committed:
Add TFResNet to doc tests

sgugger commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+        # Pad to match that done in the PyTorch Conv2D model
+        height_pad = width_pad = (self.pad_value, self.pad_value)
+        hidden_state = tf.pad(hidden_state, [(0, 0), (0, 0), height_pad, width_pad])
+        # B, C, H, W -> B, H, W, C
+        hidden_state = tf.transpose(hidden_state, (0, 2, 3, 1))
+        hidden_state = self.conv(hidden_state)
+        # B, H, W, C -> B, C, H, W
```
Those two comments would be more helpful with real names instead of capitalized initials (I know they come from PyTorch)

sgugger commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+        hidden_state = inputs
+        hidden_state = self.embedder(hidden_state)
+        hidden_state = tf.pad(hidden_state, [[0, 0], [0, 0], [1, 1], [1, 1]])
+        # B, C, H, W -> B, H, W, C
+        hidden_state = tf.transpose(hidden_state, (0, 2, 3, 1))
+        hidden_state = self.pooler(hidden_state)
+        # B, H, W, C -> B, C, H, W
```
Same here

sgugger commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+
+    def call(self, x: tf.Tensor, training: bool = False) -> tf.Tensor:
+        hidden_state = x
+        # B, C, H, W -> B, H, W, C
+        hidden_state = tf.transpose(hidden_state, (0, 2, 3, 1))
+        hidden_state = self.convolution(hidden_state)
+        # B, H, W, C -> B, C, H, W
```
And there

sgugger reviewed (approved):
Thanks for adding this model in TensorFlow! Just have a few nits.

Rocketknight1 commented:
I'm seeing a failure in `test_keras_fit` - it looks like the outputs are different depending on whether the labels are passed in the input dict or separately. That might actually be caused by some stochasticity in the model outputs, though - maybe the `training` flag isn't being passed correctly so layers like dropout are still being run in training mode during eval time?

Rocketknight1 commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+        self.config = config
+        self.embedder = TFResNetEmbeddings(config, name="embedder")
+        self.encoder = TFResNetEncoder(config, name="encoder")
+        self.pooler = TFAdaptiveAvgPool2D((1, 1), input_ordering="NCHW")
```
I just made the same comment in the RegNets PR, but a resize to `(1, 1)` should probably be replaced by the built-in [GlobalAveragePooling2D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/GlobalAveragePooling2D). I'm sorry I didn't inspect the actual calls sooner - we probably don't need that custom op in this model at all!

NielsRogge commented:
Please also incorporate the updates made in #17731

amyeroberts committed:
PR comments - GlobalAveragePooling and clearer comments

amyeroberts committed:
Remove unused import

amyeroberts committed:
Add in keepdims argument

amyeroberts commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,595 @@
+
+# Taken from:
+# https://gist.github.com/Rocketknight1/efc47242914788def0144b341b1ad638
+class TFAdaptiveAvgPool2D(tf.keras.layers.Layer):
```
I've removed this layer (re @Rocketknight1 's comment below about GlobalAveragePooling)  but will follow this pattern for the future.

Rocketknight1 commented:
Ahhh, of course! I had thought that running a single iteration of training with a learning rate of 0 would leave the weights unchanged, but that isn't true for `BatchNorm`, because `BatchNorm` weights aren't updated by gradient descent. The test was broken and we only got away with it because NLP models generally don't use `BatchNorm`. I'll fix it tomorrow!

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts commented:
@sgugger Sorry - I didn't mean to re-request for you as you'd already approved!

amyeroberts committed:
Add num_channels check

amyeroberts commented on tests/models/resnet/test_modeling_tf_resnet.py:
```diff
@@ -0,0 +1,254 @@
+    @slow
+    def test_model_from_pretrained(self):
+        for model_name in TF_RESNET_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
+            model = TFResNetModel.from_pretrained(model_name, from_pt=True)
```
To remove once all approved and weights pushed to hub

amyeroberts commented on tests/models/resnet/test_modeling_tf_resnet.py:
```diff
@@ -0,0 +1,254 @@
+    @slow
+    def test_inference_image_classification_head(self):
+        model = TFResNetForImageClassification.from_pretrained(
+            TF_RESNET_PRETRAINED_MODEL_ARCHIVE_LIST[0], from_pt=True
```
To remove once all approved and weights pushed to hub

amyeroberts commented:
@NielsRogge @Rocketknight1 friendly nudge - let me know if there's any other changes or if I'm good to merge :)

Rocketknight1 commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,491 @@
+    def call(self, x: tf.Tensor, training: bool = False) -> tf.Tensor:
+        hidden_state = x
+        # (batch_size, num_channels, height, width) -> (batch_size, height, width, num_channels)
+        hidden_state = tf.transpose(hidden_state, (0, 2, 3, 1))
+        hidden_state = self.convolution(hidden_state)
+        # (batch_size, height, width, num_channels) -> (batch_size, num_channels, height, width)
+        hidden_state = tf.transpose(hidden_state, (0, 3, 1, 2))
```
Same here, could this be an NCHW convolution instead?

amyeroberts replied:
The reason I have it this way is that Conv2D can't run on CPU with `channels_first `. You get this error: 

```
The Conv2D op currently only supports the NHWC tensor format on the CPU. The op was given the format: NCHW [Op:Conv2D]```

Rocketknight1 commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,491 @@
+
+class TFResNetStage(tf.keras.layers.Layer):
+    """
+    A ResNet stage composed by stacked layers.
```
World's nittiest nitpick:

```suggestion
    A ResNet stage composed of stacked layers.
```

Rocketknight1 reviewed (approved):
Overall looks good - my main complaint is that I think we could swap the data format for some Conv and MaxPool layers to improve performance and save memory by skipping some transposes!

ydshieh commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,491 @@
+
+    def call(self, pixel_values: tf.Tensor, training: bool = False) -> tf.Tensor:
+        _, num_channels, _, _ = shape_list(pixel_values)
+        if tf.executing_eagerly() and num_channels != self.num_channels:
+            raise ValueError(
+                "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
+            )
```
The following 2 tests (in `test_modeling_tf_core.py`) will fail

- test_saved_model_creation
- test_saved_model_creation_extended

These tests are only run for a few core models, but I think it is good to keep them working.

(I originally introduced these kinds of conditions when adding TF ViT)

Rocketknight1 replied:
Alright, that makes sense. It can stay!

amyeroberts committed:
grammar fix: by -> of

Co-authored-by: matt <rocketknight1@gmail.com>

Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>

Rocketknight1 commented on src/transformers/models/resnet/modeling_tf_resnet.py:
```diff
@@ -0,0 +1,491 @@
+        hidden_state = self.embedder(hidden_state)
+        hidden_state = tf.pad(hidden_state, [[0, 0], [0, 0], [1, 1], [1, 1]])
+        # (batch_size, num_channels, height, width) -> (batch_size, height, width, num_channels)
+        hidden_state = tf.transpose(hidden_state, (0, 2, 3, 1))
+        hidden_state = self.pooler(hidden_state)
+        # (batch_size, height, width, num_channels) -> (batch_size, num_channels, height, width)
+        hidden_state = tf.transpose(hidden_state, (0, 3, 1, 2))
```
God I hate Tensorflow

sayakpaul replied:
🤣

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Remove transposes - keep NHWC throughout forward pass

amyeroberts committed:
Merge branch 'add-tf-resnet' of github.com:amyeroberts/transformers into add-tf-resnet

amyeroberts committed:
Fixup look sharp

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Resolve conflicts after RegNet merge

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Merge branch 'main' into add-tf-resnet

amyeroberts committed:
Add missing layer names

amyeroberts committed:
Final tidy up - remove from_pt now weights on hub
